pkg/volume/glusterfs/glusterfs.go

```diff
@@ -315,3 +325,154 @@ func getVolumeSource(
+		glog.Errorf("glusterfs: failed to create gluster rest client")
+		return fmt.Errorf("glusterfs: failed to create gluster rest client, REST server authentication failed")
+	}
+	err = cli.VolumeDelete(newvolumetodel)
+	if err != nil {
+		glog.V(4).Infof("glusterfs: error when deleting the volume :%s", err)
+		return err
+	}
+	glog.V(2).Infof("glusterfs: volume %s deleted successfully", volumetodel)
+	return nil
+
+}
+
+func (r *glusterfsVolumeProvisioner) Provision() (*api.PersistentVolume, error) {
+	var err error
+	/*if r.options.ProvisionerSelector != nil {
+		glog.V(4).Infof("glusterfs: not able to parse your claim Selector")
+		return nil, fmt.Errorf("glusterfs: not able to parse your claim Selector")
+	}*/
+	glog.V(4).Infof("glusterfs: Provison VolumeOptions %v", r.options)
+	for k, v := range r.options.Parameters {
+		switch dstrings.ToLower(k) {
+		case "endpoint":
+			r.plugin.clusterconf.glusterep = v
+		case "path":
+			r.plugin.clusterconf.glusterRestvolpath = v
+		case "resturl":
+			r.plugin.clusterconf.glusterRestUrl = v
+		case "restauthenabled":
+			r.plugin.clusterconf.glusterRestAuth, err = strconv.ParseBool(v)
+		case "restuser":
+			r.plugin.clusterconf.glusterRestUser = v
+		case "restuserkey":
+			r.plugin.clusterconf.glusterRestUserKey = v
+		default:
+			return nil, fmt.Errorf("glusterfs: invalid option %q for volume plugin %s", k, r.plugin.GetPluginName())
+		}
+	}
+	glog.V(4).Infof("glusterfs: storage class parameters in plugin clusterconf %v", r.plugin.clusterconf)
+	r.glusterfsClusterConf = r.plugin.clusterconf
+	glusterfs, sizeGB, err := r.CreateVolume()
+	if err != nil {
+		glog.Errorf("glusterfs: create volume err: %s.", err)
+		return nil, fmt.Errorf("glusterfs: create volume err: %s.", err)
+	}
+	pv := new(api.PersistentVolume)
+	pv.Spec.PersistentVolumeSource.Glusterfs = glusterfs
+	pv.Spec.PersistentVolumeReclaimPolicy = r.options.PersistentVolumeReclaimPolicy
+	pv.Spec.AccessModes = r.options.AccessModes
+	pv.Spec.Capacity = api.ResourceList{
+		api.ResourceName(api.ResourceStorage): resource.MustParse(fmt.Sprintf("%dGi", sizeGB)),
+	}
+	return pv, nil
+}
+
+func (p *glusterfsVolumeProvisioner) CreateVolume() (r *api.GlusterfsVolumeSource, size int, err error) {
+	volSizeBytes := p.options.Capacity.Value()
+	sz := int(volume.RoundUpSize(volSizeBytes, 1024*1024*1024))
+	glog.V(2).Infof("glusterfs: create volume of size:%d bytes", volSizeBytes)
+	if p.glusterfsClusterConf.glusterRestUrl == "" {
+		glog.Errorf("glusterfs : rest server endpoint is empty")
+		return nil, 0, fmt.Errorf("failed to create gluster REST client, REST URL is empty")
+	}
+	cli := gcli.NewClient(p.glusterRestUrl, p.glusterRestUser, p.glusterRestUserKey)
+	if cli == nil {
+		glog.Errorf("glusterfs: failed to create gluster rest client")
+		return nil, 0, fmt.Errorf("failed to create gluster REST client, REST server authentication failed")
+	}
+	volumeReq := &gapi.VolumeCreateRequest{Size: sz}
+	volume, err := cli.VolumeCreate(volumeReq)
```

Comment:
Here volume has the ID.  Save the ID mapped to the volume.Name so that it can be retrieved later